### DISCUSSION TITLE
CLI

### DISCUSSION THREAD

**Discussion by masterofobzene** [mifi/lossless-cut#1490] [Q&A] [answered]
The page talks about "basic" cli capabilities.
I don't see any reference to the commands available or what can be done exactly with the cli.
Care to explain a little?

**Comment by mifi** [comment]
if you click on the "basic cli support" link it will take you here: https://github.com/mifi/lossless-cut/blob/master/cli.md

**Reply by masterofobzene** [reply]
Yes, I've seen it... and?... 

>  ./LosslessCut.exe arguments

What are those arguments?

**Reply by mifi** [reply]
<img width="627" alt="Screenshot 2023-02-24 at 23 43 10" src="https://user-images.githubusercontent.com/402547/221221954-4c155881-01fc-4349-bd3d-5f265073a16d.png">

**Reply by masterofobzene** [reply]
Ok... I refused to take those as "basic functionality". Thought those were simple examples.

**Reply by BazzaCuda** [reply]
After reading through the closed issues ([1204 ](https://github.com/mifi/lossless-cut/issues/1204) and [980](https://github.com/mifi/lossless-cut/issues/980)), I got the impression that you could specifiy a timecode on the command line (along with the filename) and LosslessCut would open the file with the cursor positioned at that timecode. I can't find how to formulate the command line to do that, though. Have I misunderstood? Is that capability available?
Edit: I'm running the 64-bit version on Windows 11.

**Reply by mifi** [reply]
that's not currently possible (tracked in #980)

**Reply by BazzaCuda** [reply]
ah, ok. Thanks for the reply.

**Reply by masterofobzene** [reply]
Came back to see if there is something useful to do with the CLI. 

USE CASE:

I need to cut massive quantities of videos into 30 seconds pieces each. I managed to semi-automate this with macroeditor and making an interactive macro using mouse and keyboard to interact with the GUI.



PROBLEM/LIMITATION:

I have to let the computer do the macro, this means I cannot use it while it is working. 



PROPOSED SUGGESTION:

Code a proper CLI that lets the user automate tasks. A lot of time could be saved.

**Reply by mifi** [reply]
Sounds like you're looking for [batch processing](https://github.com/mifi/lossless-cut/blob/master/batch.md)

**Reply by masterofobzene** [reply]
Can you explain how to automate the times that I need to pass to ffmpeg?
The "last command to ffmpeg" shows each command I sent which is a list of commands sent for the same file with different timestamps for each file. Yes it can do it, but we need to automate it.

I think the whole problem is that I need to pass from where to where to ffmpeg. That is why people use losslesscut instead of directly using ffmpeg.

**Reply by mifi** [reply]
There's an example by using AI along with last ffmpeg commands in the linked article. Didn't that help?
Not sure exactly what you want to automate, but if you want to automate losslessly cutting the first 30 seconds of 1000 files, the losslesscut is not the best tool for the job, and you should instead use ffmpeg. If you copy paste that sentence into e.g. chatgpt i'm sure it will provide some suggestion. if not, here's a simple script:
```bash
cd /path/to/1000-files

for f in *.mp4; do
  ffmpeg -i $f -t 30 -c copy $f-output.mp4
done
```

**Reply by masterofobzene** [reply]
Windows here. I just realized that was bash.

We need to make many videos of 30 seconds out of a big video. 
Many big videos -> many, many 30 seconds clips of each of those.

How do we pass in a programmatically way, each 30 second segment like losslesscut does?


This seems to be some kind of "uroboros". Lossless cut is an UI interface for ffmpeg. 
Then we ask for something and you reply "why?, ffmpeg can do that by doing it by hand."
*defeats the whole program existence*

Correct me if I'm wrong here, but losslesscut is an interface for ffmpeg. 
ffmpeg makes it difficult/complex to automate cutting a video in 30 second chunks because we need to tell it each segment (hard arguments) in other words from where to where. There is no magic command that we can give it to cut chunks of XX seconds.

This is were losslesscut comes in and lets us do it in a really simple way. Hurray! but we now want to do this automatically as we needed from start. 
You refuse to make a proper cli interface to it so we can interact with losslesscut.
I could write tons of lines to do it in batch, that is what I'm avoiding. In that case your program would not be required anymore since all we need is tons of time, batch (or bash in linux) and ffmpeg.

You don't seem to know how to code properly; it's ok **we are on the same boat.**
The problem is not that; **the problem is that you try to put under the rug all petitions by trying to descredit them or giving them "the manual workaround"** or in this case not working workarounds that doesn't even do what we need; but you are never sincere with your users.

A simple "I don't know how to do that" would suffice and I wouldn't / couldn't push you. We can't push you to do things beyond your capabilities. Same applies to everyone including me of course.

I will be honest, I don't know how you built this app because it is nice but then I read your answers to me and other users in the issues section and... I don't know, I'm confused.

**Reply by mifi** [reply]
> Well stop using AI, because that script is all wrong.

That script was written by me, not by AI, because I felt like giving you the benefit of the doubt and actually helping you out. You never specified that you are on windows and that you are using batch instead of the much more common bash/wsl, so that’s why I wrote it in bash (and im not as familiar with batch).

Wow this escalated. I’m the author of losslesscut so i would think i know how at least basic code. I’m just trying to help you out. The fact is that AI could have helped you translate thr script I suggested from bash to windows batch, *and* it could have helped you run the script and debug any errors. But you instead ignored my addvice and attacked me for suggesting AI to you.

You apparently haven’t even described exactly what you want to achieve yet. You just say «cut 30 seconds from each file» and thats exactly what my above script suggestion does, so I don’t know what more you expect from me.

**Reply by masterofobzene** [reply]
This is what I need:

> We need to make many videos of 30 seconds out of a big video.
> Many big videos -> many, many 30 seconds clips of each of those.

The request for a proper CLI is what I was asking. Your workarounds defeat the purpose of losslesscut. 


My comment was edited and you probably answered the old comment. I saw that bash script too similar to a "bad batch script" and given you talk wonders about AI, I simply assumed you got it from chatgpt, or deepseek. Later I realized that it was bash. And no, don't assume that wsl/bash is more common. Same for powershell (before you mention that trash of language that struggles to interact with applications and paths).


Losslesscut is awesome to do what I need. It works. It is a nice app. Congratulations. What happens to me is that I'm asking for a proper CLI interface so we can automate Losslesscut without being forced to use the gui which makes it "manual" use only.

There are petitions, I myself did one. All are disregarded as "non useful". I don't understand why would you say that other than not being able to program it. If you cannot do it, its OK. Nobody is here to pick on you or your dev capabilities. I myself learned alone and I don't know A LOT of stuff and I'm very limited. This so you know I'm not trying to bring you down.

But again, it seems that you want to hide something and return all requests for a CLI with "that is not really needed" argument.
  
Well, if its not, please tell me how to make ffmpeg cut a one hour video into segment**S** of 30 seconds. Have in mind that by doing so, it makes losslesscut obsolete.

**Reply by mifi** [reply]
Ok I see you edited your response. Now I think I see what you want. You want to losslessly split all files into 30 second chunks. This can also be done with AI but if you don’t want that then google is an alternative. I might help you write a script to do it if you can’t figure it out yourself, but you’re not really uncouraging me to do so.

The thing is that there are so many people asking for different kinds of automations, and I cannot implement all of them. That’s why I suggest scripts as a workaround. And AI because I don’t have the capacity to do free development consulting work for everybode.

**Reply by mifi** [reply]
> don't assume that wsl/bash is more common.

I would argue that bash is much more common than batch for scripting. not on windows but in total, see for example github:

- https://github.com/search?q=language%3ABatchfile&type=Repositories&ref=advsearch&l=Batchfile&l=
- https://github.com/search?q=language%3AShell&type=repositories&ref=advsearch

but yea I guess more LosslessCut users use Windows than macos+linux so for the losslesscut target group, batch is indeed probably just as common.

**Reply by mifi** [reply, marked as answer]
> Well, if its not, please tell me how to make ffmpeg cut a one hour video into segment**S** of 30 seconds. Have in mind that by doing so, it makes losslesscut obsolete.

just to prove my point that AI works I copy pasted your question:

# [ChatGTP output](https://chatgpt.com/share/679ce7ff-2458-8012-8dfb-da3665ff7494):

> please tell me how to make ffmpeg cut a one hour video into segmentS of 30 seconds

You can use `ffmpeg` to split a one-hour video into 30-second segments using the `-segment_time` option. Here’s the command:

```sh
ffmpeg -i input.mp4 -c copy -map 0 -f segment -segment_time 30 -reset_timestamps 1 output_%03d.mp4
```

### Explanation:
- `-i input.mp4` → Specifies the input video.
- `-c copy` → Copies the streams without re-encoding (fast and lossless).
- `-map 0` → Ensures all streams (video, audio, subtitles) are included in output files.
- `-f segment` → Uses the segmenting feature of `ffmpeg`.
- `-segment_time 30` → Splits the video into 30-second segments.
- `-reset_timestamps 1` → Resets timestamps in each segment to start from zero.
- `output_%03d.mp4` → Names the output files as `output_000.mp4`, `output_001.mp4`, etc.

Let me know if you need any modifications, like re-encoding or different naming formats!

> on windows

On Windows, you can run the following `ffmpeg` command in **Command Prompt (cmd)** or **PowerShell**:

```sh
ffmpeg -i input.mp4 -c copy -map 0 -f segment -segment_time 30 -reset_timestamps 1 output_%%03d.mp4
```

### Explanation:
- **`%%03d`** → On Windows, `%` needs to be escaped as `%%`, so filenames will be `output_000.mp4`, `output_001.mp4`, etc.
- Everything else works the same as in Linux/macOS.

### Steps:
1. Open **Command Prompt** (Press `Win + R`, type `cmd`, and press `Enter`).
2. Navigate to the folder containing your video using `cd` (e.g., `cd C:\path\to\video`).
3. Run the command above.

If you’re using **PowerShell**, use this version instead:

```powershell
ffmpeg -i input.mp4 -c copy -map 0 -f segment -segment_time 30 -reset_timestamps 1 output_%03d.mp4
```

Let me know if you run into any issues! 🚀

**Reply by masterofobzene** [reply]
OK I will try the suggested code in my scripts (yes, I use scripts to do stuff too and this is the only thing I was having to do "manually"). 

I didn't know ffmpeg had such functionality since in past versions you had to put the "from" and "to" timelines harcoded into the arguments, that would imply that you had to first calculate the whole first timeline from 0 to 30 secs, then "30 to 60" and so on, passing each time to ffmpeg like losslesscut does.

If it works, you might want to use it in losslesscut since when going to "last ffmpeg command" I get each line stating the hardcoded "from" and "to" times. Maybe you could simplify it using this command you are telling me, and only replace the duration that the user wants.

**Reply by mifi** [reply]
> I didn't know ffmpeg had such functionality since in past versions you had to put the "from" and "to" timelines harcoded into the arguments, that would imply that you had to first calculate the whole first timeline from 0 to 30 secs, then "30 to 60" and so on, passing each time to ffmpeg like losslesscut does.

if only you had *tried* to ask an AI in the first place, you would have known! and wouldn't have to waste your own or other people's time.

> If it works, you might want to use it in losslesscut since when going to "last ffmpeg command" I get each line stating the hardcoded "from" and "to" times. Maybe you could simplify it using this command you are telling me, and only replace the duration that the user wants.

There are many things that I could do, but like I said there are so many different things people want to use Losslesscut for and I don't have time to implement and maintain all kinds of different batch automations. LosslessCut was not built to be a batch operations toolset - It was built for super fast and efficient manual cutting. ffmpeg/bash does batching so much better.

**Reply by masterofobzene** [reply]
The line works, you just missed the  ` -c:v copy` (or ffmpeg complained about it here)

I can now use ffmpeg directly. Thank you.

**Comment by miqumi** [comment]
no way you saw ONE guy start using ai and wrote _that_ piece of text talking about masterofobzene as the entirety of 'ai opposers'... huge statistical fallacy ..

**Reply by mifi** [reply]
I'm not an AI fanboy either. But I think it can be very useful for certain tasks such as learning new command line tools or scripting. I don't like it when people just categorically boycott all AI just for the sake of it and then instead waste other people's time arguing about it, when they could instead just use AI for what it's good for. If you don't like to support US tech giants, fine, then just use a Chinese or European AI. If you care about privacy, find an AI company that also cares about it. I believe that in the future there will be an abundance of AI options available to us, many still free. We'll have to pay a price of course, but I believe there will be many good open source models that can run locally on our own computer, meaning we won't be dependent on the broligarchs or dictators, and all we'll pay for will be hardware and electricity.

**Comment by masterofobzene** [comment]
I don't understand your comment but I'm constantly using AI now for all coding tasks and I have a lot of scripts for different functions effectively reducing dependence on human developers and their apps. AI is awesome if you are not being stupid enough to use it as a chat friend or ask where to get groseries from. In fact I stopped using lossless cut. Freedom.

Yes its true that they gamble with your privacy, but if you take some precautions you can get away with really good scripts for everything.

So I was wrong, it is a great tool, you only have to be conscient of the privacy thing.
But be aware, its a matter of time until they close everything behind a paywall (I mean total commercial model, no more free use).
